enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,306 @@
+
+AWS, Azure & GCP will be added as new valid values for the `CSIDriverType` union discriminator
+in the CSIDriverConfigSpec:
+
+```go
+// CSIDriverType indicates type of CSI driver being configured.
+// +kubebuilder:validation:Enum="";vSphere
+type CSIDriverType string
+
+const (
+  AWSDriverType     CSIDriverType = "aws"
+  AzureDriverType   CSIDriverType = "azure"
+  GCPDriverType     CSIDriverType = "gcp"
+  VSphereDriverType CSIDriverType = "vSphere" // existing platform
+)
+
+// CSIDriverConfigSpec defines configuration spec that can be
+// used to optionally configure a specific CSI Driver.
+// +union
+type CSIDriverConfigSpec struct {
+  // driverType indicates type of CSI driver for which the
+  // driverConfig is being applied to.
+  //
+  // Valid values are:
+  //
+  // * aws
+  // * azure
+  // * gcp
+  // * vSphere
+  //
+  // ---
+  // Consumers should treat unknown values as a NO-OP.
+  //
+  // +kubebuilder:validation:Required
+  // +unionDiscriminator
+  DriverType CSIDriverType `json:"driverType"`
+
+  // AWS is used to configure the AWS CSI driver.
+  // +optional
+  AWS *AWSCSIDriverConfigSpec `json:"aws,omitempty"`
+
+  // Azure is used to configure the Azure CSI driver.
+  // +optional
+  Azure *AzureCSIDriverConfigSpec `json:"azure,omitempty"`
+
+  // GCP is used to configure the GCP CSI driver.
+  // +optional
+  GCP *GCPCSIDriverConfigSpec `json:"gcp,omitempty"`
+
+  ... exisiting vsphere field ... 
+}
+```
+
+ and fields for the encryption keys will be added
+to platform-specific config specs:
+
+#### AWS
+
+```go
+// AWSCSIDriverConfigSpec defines properties that can be configured for the AWS CSI driver.
+type AWSCSIDriverConfigSpec struct {
+    // KMSKeyARN sets the cluster default storage class to encrypt volumes with a user-defined KMS key,
+    // rather than the default KMS key used by AWS.
+    // According to https://docs.aws.amazon.com/service-authorization/latest/reference/list_awskeymanagementservice.html
+    // the format for a KMS key ARN is:
+    // arn:${Partition}:kms:${Region}:${Account}:key/${KeyId}
+    // +optional
+    KMSKeyARN string `json:"kmsKeyARN,omitempty"`
+}
+```
+#### Azure
+
+```go
+// AzureDiskEncryptionSet defines the configuration for a disk encryption set.
+type AzureDiskEncryptionSet struct {
+  // SubscriptionID defines the Azure subscription that contains the disk encryption set.
+  // When omitted, the subscription from the authenticating credentials of the CSI driver
+  // will be used.
+  SubscriptionID string `json:"subscriptionId"`
+  
+  // ResourceGroup defines the Azure resource group that contains the disk encryption set.
+  // When omitted, the cluster resource group from the cluster infrastructure object
+  // will be used.
+  ResourceGroup string `json:"resourceGroup"`
+  
+  // Name is the name of the disk encryption set that will be set on the default storage class.
```

Comment:
Couple of documentation nits here (need to apply this advice to other fields as well):
* The first word should be the json tag version of the field name, as opposed to the Golang field version. This is so that documentation refers to the json tag which is what the user would typically interact with
* Whenever you have a validation, you must explain it within the documentation, so here we need to add the valid characters and the maximum length
```suggestion
  // name is the name of the disk encryption set that will be set on the default storage class.
  // The value of the field consists of only of alphanumeric characters, hyphens and underscores.
  // It has a maximum length of 80 characters.
```

Reply:
@JoelSpeed Shouldn't detailed API review like this should be done when changes are being proposed to `openshift/api` ? Putting my k8s reviewer hat - it seems superfluous to do in-depth review of APIs during enhancements.

Reply:
If we have the API in detail here, it would be better that it's correct. As an API reviewer there's not a lot of difference between reviewing go embedded in markdown and the actual API PR.

If you think it's confusing to include it we can certainly move this over to the API PR, but then we are discussing the feature in two places and things could get confusing if updates aren't propagated from the API PR back to the enhancement

Reply:
If I understand the suggestion correctly here, that the first word in the comment should be the lowercase json tag, that conflicts with the example here: https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#write-user-readable-documentation-in-godoc

I can open a PR to fix the example if it is indeed incorrect.

Reply:
Yep, the example is incorrect, in this case `name` is correct, and `Name` is incorrect, I think that example was quite an early addition to the dev-guide so we may have missed including that convention that early on